### PR TITLE
types/json: support binary json type.

### PR DESCRIPTION
**Pull request by coocood**
The BinaryJSON type can directly read the elements in the serialized JSON data, we can put it in Chunk and avoid deserialization cost.

Only Marshal and Unmarshal are implemented, for now, more function will be supported later.

### CODE REVIEW

**Comment by coocood** [comment]
@hicqu @zz-jason PTAL

**Comment by shenli** [comment]
@jackysp PTAL

File: types/json/binary.go [hunk 1]
```diff
@@ -0,0 +1,601 @@
...
+}
+
+// ToJSON converts BinaryJSON to object JSON.
+func (bj BinaryJSON) ToJSON() (JSON, error) {
+	return decode(bj.TypeCode, bj.Value)
+}
+
+// MarshalJSON implements the json.Marshaler interface.
+func (bj BinaryJSON) MarshalJSON() ([]byte, error) {
+	buf := make([]byte, 0, len(bj.Value)*3/2)
```
Comment (hicqu) [inline types/json/binary.go]: `3/2` is from MySQL?

File: types/json/binary.go [hunk 2]
```diff
@@ -0,0 +1,601 @@
...
+	return bj.marshalTo(buf)
+}
+
+func (bj BinaryJSON) marshalTo(buf []byte) ([]byte, error) {
```
Comment (hicqu) [inline types/json/binary.go]: How about implement it by `visitor-mode` which is useful for implement other methods like `json_extract` or `json_set` later?

File: types/json/binary_test.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+package json
+
+import (
+	"fmt"
+
+	. "github.com/pingcap/check"
+)
+
+func (s *testJSONSuite) TestBinaryJSONMarshalUnmarshal(c *C) {
```
Comment (hicqu) [inline types/json/binary_test.go]: Add some cases to test data marshaled by old `JSON` can be unmarshaled by `BinaryJSON` success?

File: types/json/binary.go (hunk 1 above)
Comment (coocood) [reply types/json/binary.go]: No, just a wild estimate.

File: types/json/binary.go (hunk 2 above)
Comment (coocood) [reply types/json/binary.go]: `BinaryJSON` is read-only, so we only need to implement `json_extract`.
I think visitor mode is overkill.

**Comment by coocood** [comment]
@hicqu PTAL

**Comment by hicqu** [comment]
LGTM. @jackysp PTAL.

**Pull request merged**